src/main/scala/firrtl/passes/AnnotateMemMacros.scala

```diff
@@ -0,0 +1,167 @@
+import firrtl.ir._
+import firrtl._
+import firrtl.Utils._
+import firrtl.Mappers._
+
+case class AppendableInfo(fields: Map[String,Any]) extends Info {
+  def append(a: Map[String,Any]) = this.copy(fields = fields ++ a)
+  def append(a: (String,Any)): AppendableInfo = append(Map(a))
+  def get(f: String) = fields.get(f)
+  override def equals(b: Any) = b match {
+    case i: AppendableInfo => fields - "info" == i.fields - "info"
+    case _ => false
+  }
+}
+
+object AnalysisUtils {
+
+  def getConnects(m: Module) = {
+    val connects = HashMap[String, Expression]()
+    def getConnects(s: Statement): Unit = s match {
+      case s: Connect =>
+        connects(s.loc.serialize) = s.expr
+      case s: DefNode =>
+        connects(s.name) = s.value
+      case s: Block =>
+        s.stmts foreach getConnects
+      case _ =>
+    }
+    getConnects(m.body)
+    connects.toMap
+  }  
+
+  // only works in a module
+  def getConnectOrigin(connects: Map[String, Expression], node: String): Expression = {
+    if (connects contains node) getConnectOrigin(connects,connects(node))
+    else EmptyExpression
+  }
+
+  def checkLit(e: Expression) = e match {
+    case l : Literal => true
+    case _ => false
+  }
+
+  def getOrigin(connects: Map[String, Expression], e: Expression) = e match {
+    case DoPrim(_,_,_,_) => getConnectOrigin(connects,e)
+    case l if (checkLit(l)) => e
+    case _ => getConnectOrigin(connects,e.serialize)  
+  }
+
+  // takes in a list of connections in a given module and looks to find the origin of some node
+  // where origin is the source of a connect. note that if the source is a trivial primop that
+  // has yet to be optimized via constant propagation, the function will try to search backwards
+  // past the primop. this allows you to compare if two nodes have the same origin, for example. 
+  // backward searches until PrimOp, Lit or non-trivial Mux appears
+  // technically, you should keep searching through PrimOp, because a node + 0 is still itself, 
+  // a node shifted by 0 is still itself, etc. 
+  // TODO: handle validif???, more thorough
+  private def getConnectOrigin(connects: Map[String, Expression], e: Expression): Expression = e match { 
+    case Mux(cond, tv, fv, _) =>
+      val fvOrigin = getOrigin(connects,fv)
+      val tvOrigin = getOrigin(connects,tv)
+      val condOrigin = getOrigin(connects,cond)
+      if (we(tvOrigin) == we(one) && we(fvOrigin) == we(zero)) condOrigin
+      else if (we(condOrigin) == we(one)) tvOrigin
+      else if (we(condOrigin) == we(zero)) fvOrigin
+      else if (we(tvOrigin) == we(fvOrigin)) tvOrigin
+      else if (we(fvOrigin) == we(zero) && we(condOrigin) == we(tvOrigin)) condOrigin
+      else e
+    case DoPrim(PrimOps.Or, args, consts, tpe) if args.contains(one) => one
+    case DoPrim(PrimOps.And, args, consts, tpe) if args.contains(zero) => zero
+    case DoPrim(PrimOps.Bits, args, consts, tpe) =>  
+      val msb = consts(0)
+      val lsb = consts(1) 
```

Comment:
If you want you could be even more clever here and unapply consts as well. ie.
```scala
case DoPrim(PrimOps.Bits, args, Seq(msb, lsb), tpe) =>
```
You could do something similar with args.head since arg should only hvae 1 argument